### PR TITLE
RPM Fixups

### PR DESCRIPTION
**Pull request by pdevine**
Spruce up the sysconfig files and turn SELinux on by default for RPM installations.

Signed-off-by: Patrick Devine patrick.devine@docker.com

### CODE REVIEW

**Comment by jfrazelle** [comment]
could we not use systemd drop in files for this so we dont need another systemd service/socket file to maintain

ping @philips probs knows answer

**Comment by pdevine** [comment]
Yeah, I was worried about breaking other things that use systemd.  On the whole, I think having sysconfig working properly is good though.

**Comment by tianon** [comment]
This won't "work properly", though.  `EnvironmentFile` is parsed as a group of key-value pairs -- sysconfig files are sourced as shell script.

**Comment by pdevine** [comment]
@tianon I was going off of  http://fedoraproject.org/wiki/Packaging%3aSystemd  which has a whole section "EnvironmentFiles and support for /etc/sysconfig files".  EnvironmentFile should work fine here.

At the very least the RPM builds should turn on selinux, so if we don't want to reuse the sysconfig files, that's fine, but IMO the 1.7 RPM shouldn't ship with selinux turned off as that would be a regression from Red Hat's RPM.

**Comment by philips** [comment]
@jfrazelle The trouble that we keep running into is that docker cannot take configuration via environment variables and so solving this via drop-ins is pretty inelegant because @pdevine wants to do this: `/usr/bin/docker -d $DOCKER_OPTIONS $DOCKER_STORAGE_OPTIONS`.

IMHO, docker should take all of the CLI flags as environment variables, like [rakyll's globalconf][gc] does. Then we can fix this problem once and for all :)

[gc]: https://github.com/rakyll/globalconf#environment-variables

File: hack/make/.build-rpm/docker-engine.spec [hunk 1]
```diff
@@ -35,6 +35,7 @@ Requires(preun): initscripts
 
 # required packages on install
 Requires: /bin/sh
+Requires: docker-selinux
```
Comment (jfrazelle) [inline hack/make/.build-rpm/docker-engine.spec]: this installs their docker-selinux pacakge.... do we trust that it has broken our stuff in the past.

**Comment by jfrazelle** [comment]
hmmm thats a cool project and would definitely be a cool idea :)

File: hack/make/.build-rpm/docker-engine.spec (hunk 1 above)
Comment (tianon) [inline hack/make/.build-rpm/docker-engine.spec]: I'm -1 on depending on some extra SELinux-specific package that we don't
build ourselves.  If there's something that "Docker on RedHat platforms"
needs, we should include it in either Docker, or in our packages (as
appropriate).

**Comment by tianon** [comment]
I'd add that I'm not really concerned about regressions from RedHat's RPM.
The invasive patches they've previously included that ultimately we get the
bug reports for is reason enough for me (personally) to only use what
they've built as a guideline for what we ought to be doing.

We should make sure that our RPM is a good representation of how we
(upstream) officially recommend that Docker be used on a RedHat platform.
If our RPM and theirs happen to be successfully interchangeable at that
point, that's a bonus, but definitely not anywhere near a requirement from
my perspective.

From that point of view, I definitely agree that any platform which uses
SELinux ought to get the best SELinux experience we can provide out of the
box.

**Comment by jfrazelle** [comment]
@tianon +1000000

File: contrib/init/sysvinit-redhat/docker.sysconfig [hunk 2]
```diff
@@ -1,7 +1,9 @@
 # /etc/sysconfig/docker
-#
-# Other arguments to pass to the docker daemon process
-# These will be parsed by the sysv initscript and appended
-# to the arguments list passed to docker -d
 
-other_args=""
+# Modify these options if you want to change the way the docker daemon runs
+DOCKER_OPTIONS=--selinux-enabled -H fd://
```
Comment (tianon) [inline contrib/init/sysvinit-redhat/docker.sysconfig]: Why is this not just enabled by default for SELinux-enabled platforms?

**Comment by pdevine** [comment]
@tianon I definitely agree.  I pulled the docker-selinux package in because out of the box if you turn on SELinux right now the policy contexts are broken and the daemon won't even start.  I think we should be building this ourselves though.

**Comment by tianon** [comment]
@pdevine then IMO, that's a bug in Docker itself -- we need to either ship whatever bits that package installs the same way we ship AppArmor profiles, or we need to add them to our package.  What benefit does creating a separate package just for those add?  Can someone really install one without the other and have that be useful?  Why don't those artifacts live in the https://github.com/docker/docker upstream repo (under `contrib/`)?  Is SELinux support in Docker really useless without these extra bits?  (If the answer to that last one is yes, then we have a very serious problem and IMO SELinux support should be removed wholesale unless we, upstream, actually support it properly, not just via some extra package RedHat ships and upstream has no actual knowledge of).

**Comment by pdevine** [comment]
@tianon the docker-selinux package just installs the docker policy file and runs restorecon on the various docker files/directories.  I'm fairly certain if you have the selinux-policy-targeted package installed (it still has a docker.pp file in it) you can run restorecon on our stuff and things will run.  From their spec file, it looks like it's running:

``` /sbin/restorecon -R %{_bindir}/docker %{_localstatedir}/run/docker.sock %{_localstatedir}/run/docker.pid %{_sharedstatedir}/docker %{_sysconfdir}/docker %{_localstatedir}/log/docker %{_localstatedir}/log/lxc %{_localstatedir}/lock/lxc %{_usr}/lib/systemd/system/docker.service /root/.docker &> /dev/null || : \```

If you haven't done that, things will be broken out of the box.  We could potentially change our spec file to test if restorecon is present and run it, which would obviate the need for a docker-selinux package... at least until the docker.pp file is removed from selinux-policy-targeted.

**Comment by tianon** [comment]
It looks like https://bugzilla.redhat.com/show_bug.cgi?id=1200079 is a semi-related ticket, too.

I'd restate that if `docker.pp` is required for proper Docker operation with SELinux, then it ought to be checked into our repository and included with our packages, especially since using Docker+SELinux on non-RedHat platforms is entirely possible and ought to be something we're happy about people doing (and encourage people to do).

**Comment by weikinhuang** [comment]
I opened an issue where this rpm requires x86 dependencies #14042 to install. I don't know if this was intentional behavior or not since it doesn't make sense why it would use the 32bit libs when it can simply use the 64bit ones.

**Comment by jfrazelle** [comment]
closing, we dont want the docker-selinux pkg, and the others should be solved by a systemd drop in file not by specific env files because it is not consistent across distros and we do not want to maintain all these random versions

**Pull request closed without merging**